// node-poweredup/dist/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Hubs;
(function (Hubs) {
    Hubs[Hubs["UNKNOWN"] = 0] = "UNKNOWN";
    Hubs[Hubs["WEDO2_SMART_HUB"] = 1] = "WEDO2_SMART_HUB";
    Hubs[Hubs["BOOST_MOVE_HUB"] = 2] = "BOOST_MOVE_HUB";
    Hubs[Hubs["POWERED_UP_HUB"] = 3] = "POWERED_UP_HUB";
    Hubs[Hubs["POWERED_UP_REMOTE"] = 4] = "POWERED_UP_REMOTE";
    Hubs[Hubs["DUPLO_TRAIN_HUB"] = 5] = "DUPLO_TRAIN_HUB";
})(Hubs = exports.Hubs || (exports.Hubs = {}));
var Devices;
(function (Devices) {
    Devices[Devices["UNKNOWN"] = 0] = "UNKNOWN";
    Devices[Devices["BASIC_MOTOR"] = 1] = "BASIC_MOTOR";
    Devices[Devices["TRAIN_MOTOR"] = 2] = "TRAIN_MOTOR";
    Devices[Devices["LED_LIGHTS"] = 8] = "LED_LIGHTS";
    Devices[Devices["BOOST_LED"] = 22] = "BOOST_LED";
    Devices[Devices["WEDO2_TILT"] = 34] = "WEDO2_TILT";
    Devices[Devices["WEDO2_DISTANCE"] = 35] = "WEDO2_DISTANCE";
    Devices[Devices["BOOST_DISTANCE"] = 37] = "BOOST_DISTANCE";
    Devices[Devices["BOOST_TACHO_MOTOR"] = 38] = "BOOST_TACHO_MOTOR";
    Devices[Devices["BOOST_MOVE_HUB_MOTOR"] = 39] = "BOOST_MOVE_HUB_MOTOR";
    Devices[Devices["BOOST_TILT"] = 40] = "BOOST_TILT";
    Devices[Devices["DUPLO_TRAIN_BASE_MOTOR"] = 41] = "DUPLO_TRAIN_BASE_MOTOR";
    Devices[Devices["DUPLO_TRAIN_BASE_SPEAKER"] = 42] = "DUPLO_TRAIN_BASE_SPEAKER";
    Devices[Devices["DUPLO_TRAIN_BASE_COLOR"] = 43] = "DUPLO_TRAIN_BASE_COLOR";
    Devices[Devices["DUPLO_TRAIN_BASE_SPEEDOMETER"] = 44] = "DUPLO_TRAIN_BASE_SPEEDOMETER";
    Devices[Devices["POWERED_UP_REMOTE_BUTTON"] = 55] = "POWERED_UP_REMOTE_BUTTON";
})(Devices = exports.Devices || (exports.Devices = {}));
var Colors;
(function (Colors) {
    Colors[Colors["BLACK"] = 0] = "BLACK";
    Colors[Colors["PINK"] = 1] = "PINK";
    Colors[Colors["PURPLE"] = 2] = "PURPLE";
    Colors[Colors["BLUE"] = 3] = "BLUE";
    Colors[Colors["LIGHT_BLUE"] = 4] = "LIGHT_BLUE";
    Colors[Colors["CYAN"] = 5] = "CYAN";
    Colors[Colors["GREEN"] = 6] = "GREEN";
    Colors[Colors["YELLOW"] = 7] = "YELLOW";
    Colors[Colors["ORANGE"] = 8] = "ORANGE";
    Colors[Colors["RED"] = 9] = "RED";
    Colors[Colors["WHITE"] = 10] = "WHITE";
    Colors[Colors["NONE"] = 255] = "NONE";
})(Colors = exports.Colors || (exports.Colors = {}));
var ButtonStates;
(function (ButtonStates) {
    ButtonStates[ButtonStates["PRESSED"] = 0] = "PRESSED";
    ButtonStates[ButtonStates["RELEASED"] = 1] = "RELEASED";
    ButtonStates[ButtonStates["UP"] = 2] = "UP";
    ButtonStates[ButtonStates["DOWN"] = 3] = "DOWN";
    ButtonStates[ButtonStates["STOP"] = 4] = "STOP";
})(ButtonStates = exports.ButtonStates || (exports.ButtonStates = {}));
var DuploTrainBaseSounds;
(function (DuploTrainBaseSounds) {
    DuploTrainBaseSounds[DuploTrainBaseSounds["STATION_DEPARTURE"] = 5] = "STATION_DEPARTURE";
    DuploTrainBaseSounds[DuploTrainBaseSounds["WATER_REFILL"] = 7] = "WATER_REFILL";
    DuploTrainBaseSounds[DuploTrainBaseSounds["HORN"] = 9] = "HORN";
})(DuploTrainBaseSounds = exports.DuploTrainBaseSounds || (exports.DuploTrainBaseSounds = {}));
var BLEManufacturerData;
(function (BLEManufacturerData) {
    BLEManufacturerData[BLEManufacturerData["BOOST_MOVE_HUB_ID"] = 64] = "BOOST_MOVE_HUB_ID";
    BLEManufacturerData[BLEManufacturerData["POWERED_UP_HUB_ID"] = 65] = "POWERED_UP_HUB_ID";
    BLEManufacturerData[BLEManufacturerData["POWERED_UP_REMOTE_ID"] = 66] = "POWERED_UP_REMOTE_ID";
    BLEManufacturerData[BLEManufacturerData["DUPLO_TRAIN_HUB_ID"] = 32] = "DUPLO_TRAIN_HUB_ID";
})(BLEManufacturerData = exports.BLEManufacturerData || (exports.BLEManufacturerData = {}));
var BLEServices;
(function (BLEServices) {
    BLEServices["WEDO2_SMART_HUB"] = "00001523-1212-efde-1523-785feabcd123";
    BLEServices["LPF2_HUB"] = "00001623-1212-efde-1623-785feabcd123";
})(BLEServices = exports.BLEServices || (exports.BLEServices = {}));
var BLECharacteristics;
(function (BLECharacteristics) {
    BLECharacteristics["WEDO2_BATTERY"] = "2a19";
    BLECharacteristics["WEDO2_BUTTON"] = "00001526-1212-efde-1523-785feabcd123";
    BLECharacteristics["WEDO2_PORT_TYPE"] = "00001527-1212-efde-1523-785feabcd123";
    BLECharacteristics["WEDO2_LOW_VOLTAGE_ALERT"] = "00001528-1212-efde-1523-785feabcd123";
    BLECharacteristics["WEDO2_HIGH_CURRENT_ALERT"] = "00001529-1212-efde-1523-785feabcd123";
    BLECharacteristics["WEDO2_LOW_SIGNAL_ALERT"] = "0000152a-1212-efde-1523-785feabcd123";
    BLECharacteristics["WEDO2_SENSOR_VALUE"] = "00001560-1212-efde-1523-785feabcd123";
    BLECharacteristics["WEDO2_VALUE_FORMAT"] = "00001561-1212-efde-1523-785feabcd123";
    BLECharacteristics["WEDO2_PORT_TYPE_WRITE"] = "00001563-1212-efde-1523-785feabcd123";
    BLECharacteristics["WEDO2_MOTOR_VALUE_WRITE"] = "00001565-1212-efde-1523-785feabcd123";
    BLECharacteristics["WEDO2_NAME_ID"] = "00001524-1212-efde-1523-785feabcd123";
    BLECharacteristics["LPF2_ALL"] = "00001624-1212-efde-1623-785feabcd123";
})(BLECharacteristics = exports.BLECharacteristics || (exports.BLECharacteristics = {}));
